bootstrap/tracing.go

```diff
@@ -0,0 +1,155 @@
+package bootstrap
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/buildkite/agent/v3/agent"
+	"github.com/buildkite/agent/v3/env"
+	"github.com/buildkite/agent/v3/tracetools"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/exp/slices"
+	ddext "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/opentracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+// stopper lets us abstract the tracer wrap up code so we can plug in different tracing
+// library implementations that are opentracing compatible. Opentracing itself
+// doesn't have a Stop function on its Tracer interface.
+type stopper func()
+
+func noopStopper() {}
+
+func (b *Bootstrap) startTracing(ctx context.Context) (tracetools.Span, context.Context, stopper) {
+	switch b.Config.TracingBackend {
+	case tracetools.BackendDatadog:
+		// Newer versions of the tracing libs print out diagnostic info which spams the
+		// Buildkite agent logs. Disable it by default unless it's been explicitly set.
+		if _, has := os.LookupEnv("DD_TRACE_STARTUP_LOGS"); !has {
+			os.Setenv("DD_TRACE_STARTUP_LOGS", "false")
+		}
+
+		return b.startTracingDatadog(ctx)
+
+	case tracetools.BackendNone:
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+
+	default:
+		b.shell.Commentf("An invalid tracing backend was provided: %q. Tracing will not occur.", b.Config.TracingBackend)
+		b.Config.TracingBackend = "" // Ensure that we don't do any tracing after this, some of the stuff in tracetools uses the bootstrap's tracking backend
+		return &tracetools.NoopSpan{}, ctx, noopStopper
+	}
+}
+
+func (b *Bootstrap) getLabel() string {
+	label, ok := b.shell.Env.Get("BUILDKITE_LABEL")
+	if !ok {
+		label = "job"
+	}
+
+	return label
+}
+
+// startTracingDatadog sets up tracing based on the config values. It uses opentracing as an
+// abstraction so the agent can support multiple libraries if needbe.
+func (b *Bootstrap) startTracingDatadog(ctx context.Context) (tracetools.Span, context.Context, stopper) {
+	resourceName := b.OrganizationSlug + "/" + b.PipelineSlug + "/" + b.getLabel()
+	opts := []tracer.StartOption{
+		tracer.WithServiceName("buildkite_agent"),
+		tracer.WithSampler(tracer.NewAllSampler()),
+		tracer.WithAnalytics(true),
+		tracer.WithGlobalTag(ddext.ResourceName, resourceName),
+	}
+
+	tags := Merge(GenericTracingExtras(b, *b.shell.Env), DDTracingExtras())
+	opts = slices.Grow(opts, len(tags))
```

Comment:
> from [#1630](https://github.com/buildkite/agent/pull/1630#discussion_r865611722), which this PR was extracted from

review note: i'm using the new-in-Go-1.18, experimental [`slices`](https://pkg.go.dev/golang.org/x/exp/slices) package to grow our slice to the correct size here, because it's a lot clearer than using 
```Go
opts = append(opts, make([]tracer.StartOption, len(tags))...)[:len(opts)]
```

it's experimental in that the API might change, rather than experimental in that it might not work (like this PR lol), but regardless, i'm happy to replace this with non-generic code if we're not comfortable using it.

Reply:
if you're wondering why i manually grow the slice rather than just appending willy nilly (which is absolutely an option), when you call append on a slice in go, if there isn't enough capacity in the slice to add the element you want to add, it automatically doubles [0] the capacity of the slice. What this means in practice is allocating a whole other slice with the new capacity and copying everything over, which is less than ideal, but also not exactly bad.

This is fine in most cases (and honestly, probably fine here), but it can mean that we allocate more memory for a given slice than we strictly speaking need.

Because we know how big the slice is going to get ahead of time (or at least, we can derive it from the sizes of other things) it pays to allocate only as much memory to the slice.

[0] I think? I can't quite remember, but it grows it by some factor

Reply:
`slices.Grow` still uses the mechanics of `append`. It does not create a slice without more capacity than necessary: https://go.dev/play/p/cE5Fo_ydndQ

To control the capacity, one should use make and some [vanilla Go](https://go.dev/blog/slices):

```go
	tags := Merge(GenericTracingExtras(b, *b.shell.Env), DDTracingExtras())

	opts := make([]tracer.StartOption, 0, 4+len(tags))
	opts = append(opts,
		tracer.WithServiceName("buildkite_agent"),
		tracer.WithSampler(tracer.NewAllSampler()),
		tracer.WithAnalytics(true),
		tracer.WithGlobalTag(ddext.ResourceName, resourceName),
	)
	for k, v := range tags {
		opts = append(opts, tracer.WithGlobalTag(k, v))
	}
```